checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -142,6 +305,94 @@ private void visitAccess(ExpressionTree indexTree, ExpressionTree arrTree) {
     }
 
     @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+        ExecutableElement elt = TreeUtils.elementFromUse(node);
+        if (!PurityUtils.isSideEffectFree(atypeFactory, elt)) {
+            checkAnnotationsInClass(
+                    elt,
+                    null,
+                    node,
+                    getCurrentPath(),
+                    UpperBoundUtil.SideEffectKind.SIDE_EFFECTING_METHOD_CALL);
+        }
+        return super.visitMethodInvocation(node, p);
+    }
+
+    @Override
+    protected void commonAssignmentCheck(
+            Tree varTree, ExpressionTree valueExp, /*@CompilerMessageKey*/ String errorKey) {
+        super.commonAssignmentCheck(varTree, valueExp, errorKey);
+        if (varTree.getKind() == Kind.VARIABLE) {
+            // Not a reassignment, so nothing to check
+            return;
+        }
+        Element elt = InternalUtils.symbol(varTree);
+
+        if (elt != null
+                && (elt.getKind() == ElementKind.FIELD
+                        || elt.getKind() == ElementKind.PARAMETER
+                        || elt.getKind() == ElementKind.LOCAL_VARIABLE)
+                && !ElementUtils.isEffectivelyFinal(elt)) {
+            TypeKind typeKind = InternalUtils.typeOf(varTree).getKind();
+            Receiver field = FlowExpressions.internalReprOf(atypeFactory, (ExpressionTree) varTree);
+            SideEffectKind kind;
+            if (elt.getKind() == ElementKind.FIELD) {
+                if (typeKind == TypeKind.ARRAY) {
+                    kind = UpperBoundUtil.SideEffectKind.ARRAY_FIELD_REASSIGNMENT;
+                } else if (!typeKind.isPrimitive()) {
+                    kind = UpperBoundUtil.SideEffectKind.NON_ARRAY_FIELD_REASSIGNMENT;
+                } else {
+                    return;
+                }
+            } else if (typeKind == TypeKind.ARRAY) {
+                kind = SideEffectKind.LOCAL_VAR_REASSIGNMENT;
+            } else {
+                return;
+            }
+            checkAnnotationsInClass(elt, field, varTree, getCurrentPath(), kind);
```

Comment:
Why is this called for reassignments of all array types?

Reply:
Any array reassignment invalidates any annotations that have it as an argument, so we always have to check if there are any such annotations.

Reply:
If it is a public non-final field then it's reassignment can't invalidate any annotation since it can't appear in one.

Reply:
I think you're correct that I could apply this code less often - it doesn't need to be called for non-final, public fields. I thought your original comment was about the fact that local variables are handled directly above, so I'm sorry if my answer wasn't clear.

I don't think it's really necessary to add a specific case for public, non-final fields, since, as you say, they should never appear.

Reply:
Yes, at first I was just confused as to what was being checked.  Your answered clarified what was going on. 

Calling `checkAnnotationsInClass` for non-private fields won't issue any false positives, but calling it is costly.  

I think in general I'm having a hard time matching the spec we wrote in the Google Doc to the implementation.

Reply:
Maybe we should sit down and talk about this on Monday?